support/util/imagemetadata.go

```diff
@@ -131,6 +128,310 @@ func (r *RegistryClientImageMetadataProvider) ImageMetadata(ctx context.Context,
 	return config, nil
 }
 
+// GetOverride returns the image reference override based on the source/mirrors in the ICSPs/IDMSs
+func (r *RegistryClientImageMetadataProvider) GetOverride(ctx context.Context, imageRef string, pullSecret []byte) (*reference.DockerImageReference, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	var (
+		ref            *reference.DockerImageReference
+		parsedImageRef reference.DockerImageReference
+		err            error
+		overrideFound  bool
+	)
+
+	parsedImageRef, err = reference.Parse(imageRef)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse image reference %q: %w", imageRef, err)
+	}
+
+	// There are no ICSPs/IDMSs to process.
+	// That means the image reference should be pulled from the external registry
+	if len(r.OpenShiftImageRegistryOverrides) == 0 {
+		ref = &parsedImageRef
+	}
+
+	// Get the image repo info based the source/mirrors in the ICSPs/IDMSs
+	for source, mirrors := range r.OpenShiftImageRegistryOverrides {
+		for _, mirror := range mirrors {
+			ref, overrideFound, err = GetRegistryOverrides(ctx, parsedImageRef, source, mirror)
+			if err != nil {
+				log.Info(fmt.Sprintf("failed to find registry override for image reference %q with source, %s, mirror %s: %s", imageRef, source, mirror, err.Error()))
+				continue
+			}
+			break
+		}
+		// We found a successful source/mirror combo so break continuing any further source/mirror combos
+		if overrideFound {
+			break
+		}
+	}
+
+	return ref, nil
+}
+
+func (r *RegistryClientImageMetadataProvider) GetDigest(ctx context.Context, imageRef string, pullSecret []byte) (digest.Digest, *reference.DockerImageReference, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	var (
+		repo           distribution.Repository
+		ref            *reference.DockerImageReference
+		parsedImageRef reference.DockerImageReference
+		err            error
+		srcDigest      digest.Digest
+		overrideFound  bool
+	)
+
+	parsedImageRef, err = reference.Parse(imageRef)
+	if err != nil {
+		return "", nil, fmt.Errorf("failed to parse image reference %q: %w", imageRef, err)
+	}
+
+	// There are no ICSPs/IDMSs to process.
+	// That means the image reference should be pulled from the external registry
+	if len(r.OpenShiftImageRegistryOverrides) == 0 {
+		// If the image name is in the cache, return early
+		if imageDigest, exists := digestCache.Get(imageRef); exists {
+			parsedImageRef.ID = string(imageDigest.(digest.Digest))
+			return imageDigest.(digest.Digest), &parsedImageRef, nil
+		}
+
+		ref = &parsedImageRef
+	}
+
+	// Get the image repo info based the source/mirrors in the ICSPs/IDMSs
+	for source, mirrors := range r.OpenShiftImageRegistryOverrides {
```

Comment:
It looks like you run this loop in a few places, should it be in a separate function?